lib/htable.3
.\" -*- nroff -*-
.\" $Id: htable.3,v 1.2 2001/07/15 13:32:18 sandro Exp $
.Dd February 12, 2001
.Os
.Dt HTABLE 3
.Sh NAME
.Nm htable
.Nd Hash table library
.Sh SYNOPSIS
.Fd #include <htable.h>
.Bd -literal

typedef /* ... */ htable;
typedef struct {
	char *key;
	void *data;
} hpair;
typedef unsigned long (*hfunc_t)(const char *data,
                                 unsigned long table_size);

htable htable_new(void);
htable htable_new_custom(unsigned long size);
void   htable_delete(htable ht);
void   htable_set_hash_func(htable ht, hfunc_t func);
int    htable_store_key(htable ht, const char *key);
int    htable_store_data(htable ht, const char *key, void *data);
int    htable_store(htable ht, const char *key, void *data);
int    htable_exists(htable ht, const char *key);
void * htable_fetch(htable ht, const char *key);
int    htable_remove(htable ht, const char *key);
void   htable_dump(htable ht, FILE *fout);
alist  htable_list(htable ht);
.Ed
.Sh DESCRIPTION
The
.Nm
hash table library provides a generic-purpose hash table interface.
.Pp
The library provides three types:
.Fa htable ,
.Fa hpair ,
and
.Fa hfunc_t .
The
.Fa htable
type stores all the data contained into a hash table.
The
.Fa hpair
type is used by the
.Fn htable_list
function to store the hash table pairs.
The
.Fa hfunc_t
type is used by the
.Fn htable_set_hash_func
function to specify a new hashing function.
.Pp
The
.Fn htable_new
function allocates a new hash table with default size.
.Pp
The
.Fn htable_new_custom
function allocates a new hash table, with the specified size
.Fa size .
.Pp
The
.Fn htable_delete
function deallocates the previously allocated hash table
.Fa ht .
.Pp
The
.Fn htable_set_hash_func
function specifies a new hashing fuction for the hash table
.Fa ht .
.Pp
The
.Fn htable_store_key
function stores the key
.Fa key
into the hash table
.Fa ht ,
with an associated data set to
.Fa NULL .
If the key is already existing, the
.Fn htable_store_key
function will return the value -1, otherwise 0.
.Pp
The
.Fn htable_store_data
function associates the data
.Fa data
with the key
.Fa key .
If the key does not existing into the hash table, the key/data pair will not
be stored and the
.Fn htable_store_data
function will return the value -1, otherwise 0.
.Pp
The
.Fn htable_store
function stores a key/data pair into the hash table
.Fa ht .
Existing key/data pair will be overwritten.
.Pp
The
.Fn htable_exists
function returns 1 if the specified key
.Fa key
exists in the hash table
.Fa ht ,
0 otherwise.
.Pp
The
.Fn htable_fetch
function returns the data associated to the specified key
.Fa key .
If the key does not exist or no data was associated with the key, a
.Fa NULL
value will be returned.
.Pp
The
.Fn htable_remove
function removes the specified key
.Fa key
from the hash table
.Fa ht .
The
.Fn htable_remove
function returns 1 if the specified key
.Fa key
exists in the hash table
.Fa ht ,
0 otherwise.
.Pp
The
.Fn htable_dump
function dumps a user readable description of the hash table
.Fa ht
to the specified file
.Fa fout .
This is useful for debugging purposes.
.Pp
The
.Fn htable_list
function returns a new allocated (unsorted) list filled with all the pairs
(of type
.Fa hpair )
defined in the hash table.  After use, the list should be deallocated
(but not the elements, because they belong to the hash table).
.Sh EXAMPLES
Creating a hash table with some keys:
.Bd -literal -offset indent
htable ht = htable_new();
htable_store_key(ht, "key1");
htable_store_key(ht, "key2");
htable_store(ht, "key3", "key with data");
htable_store_key(ht, "key4");
htable_remove(ht, "key4");
printf("key1: %d\\n", htable_exists(ht, "key1"));
printf("key2: %d\\n", htable_exists(ht, "key2"));
printf("key3: %d\\n", htable_exists(ht, "key3"));
printf("key4: %d\\n", htable_exists(ht, "key4"));
printf("key5: %d\\n", htable_exists(ht, "key5"));
printf("key3 data: %s\\n", htable_fetch(ht, "key3"));
htable_delete(ht);
.Ed
.Pp
Dumping a hash table:
.Bd -literal -offset indent
htable ht = htable_new();
htable_store_key(ht, "key1");
htable_store_key(ht, "key2");
/* ... */
htable_dump(ht, stdout);
htable_delete(ht);
.Ed
.Sh SEE ALSO
.Xr alist 3
.Sh AUTHORS
Sandro Sigala <sandro@sigala.it>


NEWS
Changes between cutils 1.5.2 - cutils 1.6

* Cleaned up long-standing bugs and package distribution.
* chilight
	- added roff and tty formats, better handling of #xxx preprocessor
          directives and some minor changes.
          Contributed by Jukka A. Ukkonen <jau@iki.fi>.
	- added ANSI escape cleanup.
	- cleaned up HTML output.
	- added `-t' HTML title option.
	- removed useless `-d' option.
	- renamed `-s' "style" option to `-f' "format".
* cobfusc
	- cleaned up.
	- added `-A' heavy obfuscation option.
	- added `-f' separate output option.
	- added `-x' exclusive mode option.
	- added `-z' identifiers pairs option.
	- removed useless `-h' option.
	- fixed bug with the `-e' option.
	- inverted `-p' and `-r' options behavior.
* yyextract
	- cleaned up HTML output.
	- added `-t' HTML title option.
* Removed cinfo, cinfoc, cinfodc, cweave, ctangle components.

Changes between cutils 1.5.1 - cutils 1.5.2

* Bug fixes.
* `cobfusc'
	- added the digraphize option (-b).
	- simplified the trigraphize option from `-t no|all|random' to `-t'.

Changes between cutils 1.5 - cutils 1.5.1

* bug fixes

Changes between cutils 1.4 - cutils 1.5

* bug fixes
	- `chilight' ANSI escapes work again.
* added the `ctangle' utility.
* added the `cweave' utility.
* added a new Texinfo manual.

Changes between cutils 1.3.3 - cutils 1.4

* bug fixes
* added the `cunloop' utility.

Changes between cutils 1.3.2 - cutils 1.3.3

* bug fixes.
* made more portable with GNU configure.
* now compiles again under Linux.
* `cobfusc'
	- sort the dumped symbol table (-u)
* removed scc and safealloc.

Changes between cutils 1.3.1 - cutils 1.3.2

* bug fixes.

Changes between cutils 1.3 - cutils 1.3.1

* added the `configure' script to simplify the installation stage
  (idea from Masafumi NAKANE <max@FreeBSD.ORG>)
* `chilight'
	- rewritten
	- multi style support (multiple output formats).
	- HTML support (font and color).

Changes between cutils 1.2.1 - cutils 1.3

* the target OS is now expected to be FreeBSD.
* bug fixes.
	- Thanks to Liviu Daia <daia@stoilow.imar.ro>.
* added the `yyref' utility.
* added the experimental `scc' compiler.

Changes between cutils 1.2 - cutils 1.2.1

* bug fixes.
* fixed storage class bug in `cdecl'.
* removed ugly spaces in function declaration in `cdecl' and `cundecl'.

Changes between cutils 1.1 - cutils 1.2

* bug fixes.
* added the `cdecl' utility.
* added the `cundecl' utility.

Changes between cutils 1.0 - cutils 1.1

* mostly bug fixes and cleanups.
* FreeBSD compatibility fixes thanks to Tim Vanderhoek <hoek@freenet.hamilton.on.ca>.
* converted the man pages to Berkeley-style.
* added the safealloc library (snapshot).
* `yyextract'
	- added HTML format (-h) option. (nice!)
	- added width (-w) option.
* `cobfusc'
	- added dump symbol table (-u) option.
* cinfo
	- added some new stuff to the database.
	- `cinfo'
		- added HTML format (-h) option.
		- added gzipped databases support.
		- added support to references to standards.
	- `cinfodc'
		- added output headers (-h) option.
* moved files from /lib to proper directories.
* fixed syntax error messages format.

Changes between c-tools 0.4 - cutils 1.0

* all the code has been rewritten.
* covered under a BSD-like copyright.
* splitted up the sources in a clean tree.
* now is required a BSD environment to compile (BSD make, some BSD
  libc functions).
* cleaned up parameter handling.
* library:
	- cleaned up hash table functions, added more functions, added a
	  better hashing function.
* `chilight'
	- renamed from `highlight'.
	- made clean lexical analyzer `lexer.l'.
* `cobfusc'
	- renamed from `obscure'.
	- made clean lexical analyzer `lexer.l'.
	- added more functions:
		- string octalizing.
		- identifier case.
		- macro compacting.
		- symbol table handling.
		- integer garbling.
		- trigraphizing.
		- comment striping.
* cinfo stuff
	- added:
		- `cinfoc': cinfo database compiler.
		- `cinfodc': cinfo database de-compiler.
		- a cinfo database access library.
	- `cinfo'
		- new database syntax.
		- now reads the compiled database instead the database source.
	- `cinfoc'
		- new yacc parser `parser.y' and lexical analyzer `lexer.l'.
* added `yyextract'.
* removed `hsort'.


lib/htable.c
/*	$Id: htable.c,v 1.1 2001/07/13 18:28:45 sandro Exp $	*/

/*
 * Copyright (c) 1997-2001 Sandro Sigala.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifdef TEST
#undef NDEBUG
#endif
#include <assert.h>
#include <stdlib.h>
#include <string.h>

#include "config.h"
#include "htable.h"

#define HTABLE_DEFAULT_SIZE		127

typedef struct hbucket_s *hbucket;

struct hbucket_s {
	hpair	pair;
	hbucket	next;
};

struct htable_s {
	unsigned long size;
	hbucket *table;
	hfunc_t hfunc;
};

/*
 * Default hashing function.
 */
static unsigned long htable_hash(const char *p, unsigned long size)
{
	unsigned long hash = 0, g;

	while (*p != '\0') {
		hash = (hash << 4) + *p++;
		if ((g = hash & 0xF0000000) != 0)
			hash ^= g >> 24;
		hash &= ~g;
	}

	return hash % size;
}

static hbucket build_bucket(const char *key)
{
	hbucket ptr;

	ptr = (hbucket)xmalloc(sizeof(*ptr));
	memset(ptr, 0, sizeof(*ptr));
	ptr->pair.key = xstrdup(key);

	return ptr;
}

static void free_bucket(hbucket ptr)
{
	if (ptr->pair.key != NULL)
		free(ptr->pair.key);

	free(ptr);
}

htable htable_new(void)
{
	return htable_new_custom(HTABLE_DEFAULT_SIZE);
}

htable htable_new_custom(unsigned long size)
{
	htable ptr;

	ptr = (htable)xmalloc(sizeof(*ptr));
	ptr->size = size;
	ptr->table = (hbucket *)xmalloc(ptr->size * sizeof(hbucket));
	memset(ptr->table, 0, ptr->size * sizeof(hbucket));
	ptr->hfunc = htable_hash;

	return ptr;
}

void htable_delete(htable ht)
{
	hbucket bucket, next;
	unsigned int i;

	for (i = 0; i < ht->size; i++) {
		bucket = ht->table[i];
		while (bucket != NULL) {
			next = bucket->next;
			free_bucket(bucket);
			bucket = next;
		}
	}

	free(ht->table);
	free(ht);
}

void htable_set_hash_func(htable ht, hfunc_t hfunc)
{
	ht->hfunc = hfunc;
}

int htable_store_key(htable ht, const char *key)
{
	unsigned long hash;

	if (htable_exists(ht, key))
		return -1;

	hash = ht->hfunc(key, ht->size);

	if (ht->table[hash] == NULL)
		ht->table[hash] = build_bucket(key);
	else {
		hbucket bucket = build_bucket(key);
		bucket->next = ht->table[hash];
		ht->table[hash] = bucket;
	}

	return 0;
}

int htable_store_data(htable ht, const char *key, void *data)
{
	hbucket bucket;
	unsigned long hash;

	hash = ht->hfunc(key, ht->size);

	for (bucket = ht->table[hash]; bucket != NULL; bucket = bucket->next)
		if (!strcmp(bucket->pair.key, key)) {
			bucket->pair.data = data;
			return 0;
		}

	return -1;
}

int htable_store(htable ht, const char *key, void *data)
{
	if (htable_exists(ht, key))
		htable_remove(ht, key);

	htable_store_key(ht, key);
	htable_store_data(ht, key, data);

	return 0;
}

int htable_exists(htable ht, const char *key)
{
	hbucket bucket;
	unsigned long hash;

	hash = ht->hfunc(key, ht->size);

	for (bucket = ht->table[hash]; bucket != NULL; bucket = bucket->next)
		if (!strcmp(bucket->pair.key, key))
			return 1;

	return 0;
}

void *htable_fetch(htable ht, const char *key)
{
	hbucket bucket;
	unsigned long hash;

	hash = ht->hfunc(key, ht->size);

	for (bucket = ht->table[hash]; bucket != NULL; bucket = bucket->next)
		if (!strcmp(bucket->pair.key, key))
			return bucket->pair.data;

	return NULL;
}

int htable_remove(htable ht, const char *key)
{
	hbucket bucket, last = NULL;
	unsigned long hash;

	hash = ht->hfunc(key, ht->size);

	for (bucket = ht->table[hash]; bucket != NULL; bucket = bucket->next) {
		if (!strcmp(bucket->pair.key, key)) {
			if (bucket == ht->table[hash])
				ht->table[hash] = bucket->next;
			else
				last->next = bucket->next;
			free_bucket(bucket);
			return 0;
		}
		last = bucket;
	}

	return -1;
}

void htable_dump(htable ht, FILE *fout)
{
	hbucket bucket;
	unsigned int i;
	int nelements = 0, dups = 0;

	fprintf(fout, "Hash table dump:\n");
	for (i = 0; i < ht->size; ++i) {
		bucket = ht->table[i];
		if (bucket != NULL) {
			fprintf(fout, "table[%d] = {\n", i);
			for (; bucket != NULL; bucket = bucket->next) {
				++nelements;
				if (bucket != ht->table[i])
					++dups;
				fprintf(fout, "\ttable[\"%s\"] = %p\n",
					bucket->pair.key, bucket->pair.data);
			}
			fprintf(fout, "}\n");
		}
	}
	fprintf(fout, "Table size: %ld\n", ht->size);
	fprintf(fout, "Number of elements: %d\n", nelements);
	fprintf(fout, "Duplicated hash values: %d\n", dups);
	fprintf(fout, "Performance: %.2f%%\n", (1-(float)dups/nelements)*100);
}

alist htable_list(htable ht)
{
	alist al;
	hbucket bucket;
	unsigned int i;

	al = alist_new();
	for (i = 0; i < ht->size; ++i)
		if ((bucket = ht->table[i]) != NULL)
			for (; bucket != NULL; bucket = bucket->next)
				alist_append(al, &bucket->pair);
	return al;
}

#ifdef TEST

int main(void)
{
	htable ht = htable_new();

	assert(htable_store_key(ht, "hello") == 0);
	assert(htable_store_key(ht, "world") == 0);
	assert(htable_store_key(ht, "bar") == 0);
	assert(htable_store_key(ht, "hello") != 0);
	assert(htable_exists(ht, "hello"));
	assert(htable_exists(ht, "world"));
	assert(htable_remove(ht, "world") == 0);
	assert(!htable_exists(ht, "world"));
	assert(htable_store_key(ht, "world") == 0);
	assert(htable_store_data(ht, "hello", "hello data") == 0);
	assert(htable_store_data(ht, "baz", "baz data") != 0);
	assert(!htable_exists(ht, "baz"));
	assert(htable_store(ht, "foo", "foo data") == 0);
	assert(htable_store_key(ht, "var1") == 0);
	assert(htable_store_key(ht, "var2") == 0);
	assert(htable_fetch(ht, "foo") != NULL && 
	       !strcmp(htable_fetch(ht, "foo"), "foo data"));
	assert(htable_fetch(ht, "hello") != NULL && 
	       !strcmp(htable_fetch(ht, "hello"), "hello data"));
	htable_dump(ht, stdout);
	htable_delete(ht);
	printf("htable test successful.\n");

	return 0;
}

#endif /* TEST */
